pkg/store/store.go

```diff
@@ -175,12 +176,57 @@ func (s *atomixStore) List(ctx context.Context, filters *topoapi.Filters) ([]top
 
 	eps := make([]topoapi.Object, 0)
 
-	for entry := range mapCh {
-		if ep, err := decodeObject(entry); err == nil {
-			if match(ep, filters) {
+	// first make sure there are filters. if there aren't, just return everything
+	if filters == nil {
+		for entry := range mapCh {
+			if ep, err := decodeObject(entry); err == nil {
 				eps = append(eps, *ep)
 			}
 		}
+		return eps, nil
+	}
+
+	if filters.RelationFilter != nil {
+		filter := filters.RelationFilter
+
+		// contains _all_ relations that have the same kind as the filter, same SrcId as the filter, and has a target ID that (at time of their reading in mapCh) had not been seen
+		entitiesToGet := make(map[topoapi.ID]*topoapi.Object)
+
+		for entry := range mapCh {
+			if ep, err := decodeObject(entry); err == nil {
+				// if object is a relation and its kind and src id matches the filter, push the destination id
+				if ep.Type == topoapi.Object_RELATION && string(ep.GetRelation().KindID) == filter.GetRelationKind() && string(ep.GetRelation().GetSrcEntityID()) == filter.SrcId {
+					entitiesToGet[ep.GetRelation().TgtEntityID] = nil
+				} else
+				// if object is an entity, see if satisfies some relation. else, put in unresolved_entities
+				if ep.Type == topoapi.Object_ENTITY {
+					if map_entity, found := entitiesToGet[ep.ID]; found {
+						if filter.TargetKind == "" || ep.GetKind().Name == filter.TargetKind {
+							*map_entity = *ep
+						}
+					}
+				}
+			}
+		}
+		// iterate over entities to make sure we did not miss any valid ones (due to the corresponding relationship being seen first)
+		for id := range entitiesToGet {
```

Comment:
You can do `for id, entity := range ....` then you can just look at entity == nil instead of map test.